From d800ede1080c389a7293ccf4ad5d4b5c443f0995 Mon Sep 17 00:00:00 2001
From: Kevin McCarthy <kevin@8t8.us>
Date: Wed, 19 Feb 2020 19:30:41 -0800
Subject: [PATCH] Pass sctx by reference to mutt_send_message_resume().

This makes it more noticable that the parameter may be freed, and also
prevents a use after free.
---
 background.c |  4 ++--
 send.c       | 13 +++++++++----
 send.h       |  2 +-
 3 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/background.c b/background.c
index ef27a9f8..c6b5e6cf 100644
--- a/background.c
+++ b/background.c
@@ -504,7 +504,7 @@ void mutt_background_compose_menu (void)
     sctx = process->sctx;
     process_list_remove (process);
     bg_process_free (&process);
-    if (mutt_send_message_resume (sctx) == 2)
+    if (mutt_send_message_resume (&sctx) == 2)
     {
       /* L10N:
          Message displayed when the user chooses to background
@@ -550,7 +550,7 @@ void mutt_background_compose_menu (void)
           process_list_remove (process);
           bg_process_free (&process);
 
-          if (mutt_send_message_resume (sctx) == 2)
+          if (mutt_send_message_resume (&sctx) == 2)
           {
             /* L10N:
                Message displayed when the user chooses to background
diff --git a/send.c b/send.c
index b380b6a0..8b1072b5 100644
--- a/send.c
+++ b/send.c
@@ -2531,16 +2531,21 @@ cleanup:
 
 /* backgroundable and resumable part of the send process.
  *
- * need to define a "backgrounded" return value.
+ * *psctx will be freed unless the message is backgrounded again.
  *
  * Returns 0 if the message was successfully sent
  *        -1 if the message was aborted or an error occurred
  *         1 if the message was postponed
  *         2 if the message editing was backgrounded
  */
-int mutt_send_message_resume (SEND_CONTEXT *sctx)
+int mutt_send_message_resume (SEND_CONTEXT **psctx)
 {
   int rv;
+  SEND_CONTEXT *sctx;
+
+  if (!psctx || !*psctx)
+    return -1;
+  sctx = *psctx;
 
   if (sctx->local_scope)
   {
@@ -2568,7 +2573,7 @@ cleanup:
   }
 
   if (rv != 2)
-    send_ctx_free (&sctx);
+    send_ctx_free (psctx);
 
   return rv;
 }
@@ -2627,7 +2632,7 @@ mutt_send_message (int flags,            /* send mode */
    * the sctx if appropriate, and also adds to the background edit
    * list.
    */
-  rv = mutt_send_message_resume (sctx);
+  rv = mutt_send_message_resume (&sctx);
   if (rv == 2)
   {
     /* TODO:
diff --git a/send.h b/send.h
index 05be1aa5..fbc38bb1 100644
--- a/send.h
+++ b/send.h
@@ -94,6 +94,6 @@ ADDRESS *mutt_default_from (void);
 void mutt_encode_descriptions (BODY *, short);
 int mutt_resend_message (FILE *, CONTEXT *, HEADER *);
 int mutt_send_message (int, HEADER *, const char *, CONTEXT *, HEADER *);
-int mutt_send_message_resume (SEND_CONTEXT *sctx);
+int mutt_send_message_resume (SEND_CONTEXT **psctx);
 
 #endif
-- 
2.26.2

